### DISCUSSION TITLE
Soundness of BufferViewMut

### DISCUSSION THREAD

**Discussion by agausmann** [gfx-rs/wgpu#2582] [Q&A] [answered]
Although attempting to get a `&[u8]` from a write-only `BufferViewMut` [will panic](https://docs.rs/wgpu/0.12.0/src/wgpu/lib.rs.html#2042-2053), getting a `&mut [u8]` is considered perfectly fine, which seems suspicious to me, because `&mut [u8]` is still readable.

If we try to perform a read, what is going to happen, is the behavior defined? And even if it is defined soundly, if it crashes the program (e.g. it generates an access violation), I'd still consider it a hazardous interface because we can't guarantee that the compiler won't emit a `load`, even if we never explicitly access it in the source code.

**Comment by cwfitzgerald** [comment, marked as answer]
It's perfectly defined, it's just going to be potentially slow as hell due to the kind of caching behavior inherent with write-optimized mappings. The whole "getting a read only view panics" is just an attempt to discourage reading from it, as ultimately rust has no concept of a "write only" type.

**Reply by agausmann** [reply]
I forgot to mention, the reason why I mention and assume that the view is write-only is because that's implied in the documentation for [`MapMode::Write`](https://docs.rs/wgpu/0.12.0/wgpu/enum.MapMode.html).

But based on your answer, I understand it's just heavily write-optimized and not intended to be used for reading, however reads are still allowed. Would that be a better summary?

**Reply by agausmann** [reply]
And I'd be curious to know, are there any coherency guarantees about reading the same value after writing?

**Reply by cwfitzgerald** [reply]
> I understand it's just heavily write-optimized and not intended to be used for reading, however reads are still allowed. Would that be a better summary?

Correct.

> are there any coherency guarantees about reading the same value after writing?

I believe it will be the same coherence as any other store.
